### ISSUE TITLE
Add DBS migration integration test which includes multiple parents

### ISSUE THREAD

**Issue by vkuznet** [dmwm/dbs2go#58]
While working on https://github.com/dmwm/dbs2go/issues/53, and resolving https://github.com/dmwm/CRABServer/issues/7315 and https://github.com/dmwm/CRABServer/issues/7290 I came across the following use-case:
- for AODSIM dataset we have 2 levels of parentage, GEN-SIM-RAW and GEN-SIM
- the migration server should properly define order of injection, i.e. original dataset (the lowest parent) to all children, in this case the chain of injection should following GEN-SIM -> GEN-SIM-RAW -> AODSIM
- for testing purposes I used the following migration request
```
{
    "migration_url": "https://cmsweb-testbed.cern.ch/dbs2go",
    "migration_input": "/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/AODSIM"
}
```
To perform the work I started DSBMigrate and DBSMigration services as following:
```
# DBS Migrate server
./dbs2go -config migrate.json

# DBSMigration server
./dbs2go -config migration.json
```
and, I injected the data as following:
```
curl -H "Content-Type: application/json" -d@$PWD/d.json  http://localhost:9898/dbs2go-migrate/submit
```
In total it yields 10 blocks to be injected.

Once the process is started I can verify the migration status  as following:
```
curl http://localhost:9898/dbs2go-migrate/status > log
```
And, finally, I can check the status of individual migration requests as following:
```
cat log | jq '.[] | [.migration_request_id,.migration_status] | @csv'
```
Each individual migration request has the following status:
```
{"create_by":"DBS-workflow",
"creation_date":1656683373,
"last_modification_date":1656683373,
"last_modified_by":"DBS-workflow",
"migration_input":"/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/AODSIM#3517e1b6-76e3-11e7-a0c8-02163e00d7b3",
"migration_request_id":9,
"migration_status":2,
"migration_url":"https://cmsweb-testbed.cern.ch/dbs2go","retry_count":0}
```
The migration statuses are defined here https://github.com/dmwm/dbs2go/blob/master/dbs/migrate.go#L44-L67 and status 2 corresponds to success migration. Therefore, with this command:
```
cat log | jq '.[] | [.migration_request_id,.migration_status] | @csv'
"1,2"
"2,2"
"3,2"
"4,2"
"5,2"
"6,2"
"7,2"
"8,2"
"9,2"
"10,2"
```
I can easily see that all migrations were successful. 

So, we need to integrate this procedure to the integration test and verify that all migration are successful.

**Comment by amaltaro** [comment]
> for AODSIM dataset we have 2 levels of parentage, GEN-SIM-RAW and GEN-SIM

this is a surprise to me! I always thought that a dataset/file can only have 1 parent dataset (multiple blocks and/or multiple files within that SAME dataset are of course expected).

**Comment by vkuznet** [comment]
Alan, this is what I meant, the AODSIM depends on GEN-SIM-RAW, and GEN-SIM-RAW depends on GEN-SIM. So it is two levels of parents we need to migrate (migration process should migrate all parents regardless how deep they are), but it is one parent per dataset.

**Comment by amaltaro** [comment]
Ahh, understood! This is only meaningful for dataset migration across DBS instances. Sorry, I confused it with data parentage in DBS.

**Comment by vkuznet** [comment]
@d-ylee , Dennis I suggest to create series of json files which will have different migration requests, and add a loop over this files in integration test. The loop should only look-up files either by pattern or look them up from specific test/data/migration_requests area. Doing this way the test will be independent from number of files it should process and we can add more and more examples (json files). Please follow this logic when you'll develop this test.

**Comment by d-ylee** [comment]
@vkuznet To clarify my understanding, I should first inject unique bulkblocks data with DBSWriter with the corresponding datasets and parent datasets (GEN-SIM -> GEN-SIM-RAW -> AODSIM), then do the migration request on the lowest child dataset?

**Comment by vkuznet** [comment]
Now, you do not need to do that. Instead please use DBSMigrate APIs to submit different migration requests.
Let's say you have migration request in this file:
```
cat d.txt
{
    "migration_url": "https://cmsweb.cern.ch/dbs/prod/global/DBSReader",
    "migration_input": "/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/AODSIM"
}
```
It means that the migration server will use `migration_url` to get the data. the `migration_input` provides the dataset or block in your request. In this case, the AODSIM dataset has parent `/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/GEN-SIM-RAW` which by itself has parent `/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/GEN-SIM`. As such all parents should be migrated when you specify AODSIM dataset. You can submit this request to migration server as dollowing:
```
curl -H "Content-Type: application/json" http://localhost:9898/dbs2go-migrate/submit -d@$PWD/d.json
```
where URL will be the one where you run DBS Migrate server. But the migration will happen only if you have both DBSMigrate and DBS migration server. The first one, provides APIs to submit, remove, or get status of migration requests, while later is a daemon who process migration requests. Please refer to migration server [documentation](https://github.com/dmwm/dbs2go/blob/master/docs/MigrationServer.md).

I ask you to write integration test which will do migration request from series of files:
- create new area where we will store json files (like the on I shown above)
- create unit test which will read these files (regardless of number of them) and call `/submit` API
- you'll need to add `/status` API call to check migration status of your requests too
- you'll need to add time sleep to wait for all migration requests to be processed
- you'll need to start DBSMigrate and DBS migration servers too and then shut them down after requests are done
- your status call should either have `"migration_status":2` (successful migration code) or `"migration_status":9` (terminated failure)

Here is a full example how I test migration requests:
- in terminal 1 start `./dbs2go -config migrate.json` where `migrate.json` file is the following
```
{
    "base": "/dbs2go-migrate",
    "port": 9898,
    "staticdir": "/Users/vk/Work/Languages/Go/gopath/src/github.com/dmwm/dbs2go/static",
    "graphqlSchema": "/Users/vk/Work/Languages/Go/gopath/src/github.com/dmwm/dbs2go/static/schema/schema.graphql",
    "dbfile": "./dbfile.tmp",
    "migration_dbfile": "./dbmigration.tmp",
    "server_type": "DBSMigrate",
    "lexicon_file": "./static/lexicon_writer.json",
    "api_parameters_file": "./static/parameters.json",
    "concurrent_bulkblocks": true,
    "servercrt": "",
    "serverkey": "",
    "hkey": "",
    "updateDNs": 60,
    "log_file": "/tmp/migrate.log",
    "verbose": 2
}
```
- in second terminal start `./dbs2go -config migration.json` where `migration.json` file is the following
```
{
    "base": "/dbs2go",
    "port": 9899,
    "staticdir": "/Users/vk/Work/Languages/Go/gopath/src/github.com/dmwm/dbs2go/static",
    "graphqlSchema": "/Users/vk/Work/Languages/Go/gopath/src/github.com/dmwm/dbs2go/static/schema/schema.graphql",
    "dbfile": "./dbfile.tmp",
    "migration_dbfile": "./dbmigration.tmp",
    "server_type": "DBSMigration",
    "lexicon_file": "./static/lexicon_writer.json",
    "concurrent_bulkblocks": true,
    "servercrt": "",
    "serverkey": "",
    "hkey": "",
    "updateDNs": 60,
    "log_file": "/tmp/migration.log",
    "verbose": 1
}
```
- in third terminal I make the following calls:
  - submit migration request `curl -H "Content-Type: application/json" http://localhost:9898/dbs2go-migrate/submit -d@$PWD/d.json` where `d.json` I shown above
  - check migration status `curl http://localhost:9898/dbs2go-migrate/status`
  - optionally I check migration statuses as
     - `curl http://localhost:9898/dbs2go-migrate/status > log`
     - `cat log | jq '.[] | [.migration_request_id,.migration_status] | @csv'`

All of the steps above you'll do in goroutines, i.e. you can start migrate and migration services, you can make API call (you don't need curl for that just use go HTTP library), and you can check status. For instance, for `d.json` (AODSIM dataset) we got the following status:
```
[
{"create_by":"DBS-workflow","creation_date":1657640833,"last_modification_date":1657640833,"last_modified_by":"DBS-workflow","migration_input":"/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/GEN-SIM#34dcb376-76b5-11e7-a0c8-02163e00d7b3","migration_request_id":1,"migration_status":0,"migration_url":"https://cmsweb-testbed.cern.ch/dbs2go","retry_count":0}
,{"create_by":"DBS-workflow","creation_date":1657640833,"last_modification_date":1657640833,"last_modified_by":"DBS-workflow","migration_input":"/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/GEN-SIM#7b7ce6a4-75e1-11e7-a0c8-02163e00d7b3","migration_request_id":2,"migration_status":0,"migration_url":"https://cmsweb-testbed.cern.ch/dbs2go","retry_count":0}
,{"create_by":"DBS-workflow","creation_date":1657640833,"last_modification_date":1657640833,"last_modified_by":"DBS-workflow","migration_input":"/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/GEN-SIM#b6ec42ec-76b4-11e7-a0c8-02163e00d7b3","migration_request_id":3,"migration_status":0,"migration_url":"https://cmsweb-testbed.cern.ch/dbs2go","retry_count":0}
,{"create_by":"DBS-workflow","creation_date":1657640833,"last_modification_date":1657640833,"last_modified_by":"DBS-workflow","migration_input":"/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/GEN-SIM#778ff06c-75f1-11e7-a0c8-02163e00d7b3","migration_request_id":4,"migration_status":0,"migration_url":"https://cmsweb-testbed.cern.ch/dbs2go","retry_count":0}
,{"create_by":"DBS-workflow","creation_date":1657640833,"last_modification_date":1657640833,"last_modified_by":"DBS-workflow","migration_input":"/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/GEN-SIM-RAW#70fe3a1a-768c-11e7-a0c8-02163e00d7b3","migration_request_id":5,"migration_status":0,"migration_url":"https://cmsweb-testbed.cern.ch/dbs2go","retry_count":0}
,{"create_by":"DBS-workflow","creation_date":1657640833,"last_modification_date":1657640833,"last_modified_by":"DBS-workflow","migration_input":"/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/GEN-SIM-RAW#eee377dc-76e2-11e7-a0c8-02163e00d7b3","migration_request_id":6,"migration_status":0,"migration_url":"https://cmsweb-testbed.cern.ch/dbs2go","retry_count":0}
,{"create_by":"DBS-workflow","creation_date":1657640833,"last_modification_date":1657640833,"last_modified_by":"DBS-workflow","migration_input":"/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/GEN-SIM-RAW#70f88b60-768c-11e7-a0c8-02163e00d7b3","migration_request_id":7,"migration_status":0,"migration_url":"https://cmsweb-testbed.cern.ch/dbs2go","retry_count":0}
,{"create_by":"DBS-workflow","creation_date":1657640833,"last_modification_date":1657640833,"last_modified_by":"DBS-workflow","migration_input":"/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/GEN-SIM-RAW#eee338e4-76e2-11e7-a0c8-02163e00d7b3","migration_request_id":8,"migration_status":0,"migration_url":"https://cmsweb-testbed.cern.ch/dbs2go","retry_count":0}
,{"create_by":"DBS-workflow","creation_date":1657640833,"last_modification_date":1657640833,"last_modified_by":"DBS-workflow","migration_input":"/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/AODSIM#3517e1b6-76e3-11e7-a0c8-02163e00d7b3","migration_request_id":9,"migration_status":0,"migration_url":"https://cmsweb-testbed.cern.ch/dbs2go","retry_count":0}
,{"create_by":"DBS-workflow","creation_date":1657640833,"last_modification_date":1657640833,"last_modified_by":"DBS-workflow","migration_input":"/GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/AODSIM#35197562-76e3-11e7-a0c8-02163e00d7b3","migration_request_id":10,"migration_status":0,"migration_url":"https://cmsweb-testbed.cern.ch/dbs2go","retry_count":0}
]

```
So, in total it is 10 blocks which will be migrated.

Please try it out within terminal setup first to get better understand about migration request flow. Then you can write proper code to simulate this.

The only caveat is how to stop checking status. But for that, as I wrote, we should have either successful code, or failed one. Please note, the failure code is 3, but migration service will try 3 times, and after that it will switch code to 9. And, if we will test PRODUCTION datasets the code will be 4. Therefore, the test should check codes 2, 4, 9, which are final ones when migration is either done or failed. All other codes are intermediate. Please refer to [code](https://github.com/dmwm/dbs2go/blob/master/dbs/migrate.go) and [documentation](https://github.com/dmwm/dbs2go/blob/master/docs/MigrationServer.md)

**Comment by vkuznet** [comment]
Dennis, not necessarily. We can create first s snapshot of SQLite db and use this file. For instance, we can first migrate all interested datasets from global DBS to local SQLite. then add such file to repository and use it in local tests. For that one we will need to start DBS reader with that DB and it will be used instead of DBS global.

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by vkuznet** [comment]
I suggest to add the following block `/SingleMuon/Run2017E-UL2017_MiniAODv2-v1/MINIAOD#1d522d75-7490-4fff-9a21-198abe0b2bb9` to the list we should test during validation. This block has diamond-like parentage, i.e. initial MINIAOD dataset/block depends on many AOD ones, where individual AOD depend on a set of RAW datasets/blocks, and some have common RAW dataset/blocks.

**Comment by vkuznet** [comment]
@d-ylee where do we stand on this issue? We are planning to upgrade DBSMigration/DBSMigrate servers on production nodes and it would be very useful to have in place unit/integration tests that we can easily test new releases before the upgrade procedure on k8s infrastructure.

**Comment by d-ylee** [comment]
@vkuznet Sorry about the late responses. I have been tied up with projects that I have recently been added to. I submitted a draft pull request to show you an approach that I came up with your previous guidance. Please let me know your thoughts.

One other thing. I tried to do a request with the `/SingleMuon/Run2017E-UL2017_MiniAODv2-v1/MINIAOD#1d522d75-7490-4fff-9a21-198abe0b2bb9` block, but it seemed like the request took a very long time to complete. Even after one day, when I tried to get the migration `/status`, it returns error about database lock. What would you suggest for this case?

**Comment by vkuznet** [comment]
Some datasets can be very large, e.g. recently I found a dataset with 7K parent blocks. And on my laptop it took hours to migrate. We can't use those in unit tests which should complete fast enough. But DB lock issue should be investigated. I saw it before I switched to async DBSMigrate logic, and I fixed the issue in a code which didn't use transaction. Therefore, I cannot say it this issue is there or not. I think we will need to see it once you'll complete stand-along unit test logic outlined in my comment in https://github.com/dmwm/dbs2go/pull/75
